File: 0992-subarrays-with-k-different-integers/0992-subarrays-with-k-different-integers.java
class Solution {
    public int slidingWindow(int[] nums, int k) {
        Map<Integer, Integer> mp = new HashMap<>();
        int i = 0, j = 0, count = 0, n = nums.length;
        while (j < n) {
            mp.put(nums[j], mp.getOrDefault(nums[j], 0) + 1);
            while (mp.size() > k) {
                mp.put(nums[i], mp.getOrDefault(nums[i], 0) - 1);
                if (mp.get(nums[i]) == 0) {
                    mp.remove(nums[i]);
                }
                i++;
            }
            count += j - i + 1;
            j++;
        }
        return count;
    }
    public int subarraysWithKDistinct(int[] nums, int k) {
        return slidingWindow(nums, k) - slidingWindow(nums, k - 1);
    }
}

File: LeetCode/0992-subarrays-with-k-different-integers/0992-subarrays-with-k-different-integers.py
class Solution:
    def slidingWindow(self, nums: List[int], k: int) -> int:
        mp = defaultdict(int)
        i, count, n = 0, 0, len(nums)
        for j in range(n):
            mp[nums[j]] += 1
            while len(mp) > k:
                mp[nums[i]] -= 1
                if mp[nums[i]] == 0:
                    del mp[nums[i]]
                i += 1
            count += j - i + 1
        return count

    def subarraysWithKDistinct(self, nums: List[int], k: int) -> int:
        return self.slidingWindow(nums, k) - self.slidingWindow(nums, k - 1)